gwctl/pkg/common/clients.go

```diff
@@ -118,9 +119,52 @@ func MustClientsForTest(t *testing.T, initRuntimeObjects ...runtime.Object) *K8s
 		WithIndex(&corev1.Event{}, "involvedObject.namespace", eventNamespaceExtractorFunc).
 		WithIndex(&corev1.Event{}, "involvedObject.uid", eventUIDExtractorFunc).
 		Build()
-	fakeDC := fakedynamicclient.NewSimpleDynamicClient(scheme, initRuntimeObjects...)
 	fakeDiscoveryClient := fakeclientset.NewSimpleClientset().Discovery()
 
+	// Setup a fake DynamicClient, which requires some special handling.
+	//
+	// When objects are injected using `NewSimpleDynamicClient` or
+	// `NewSimpleDynamicClientWithCustomListKinds` to the fake resource tracker,
+	// internally it will use the `meta.UnsafeGuessKindToResource()` function.
+	// This incorrectly guesses the GVR of Gateway with Resource being guessed as
+	// "gatewaies" (instead of "gateways"). As a workaround, we will have to
+	// create Gateway objects separately.
+	//
+	// Also, because of this incorrect guessing, we need to register the
+	// GatewayList type separately.
+	gatewayv1GVR := schema.GroupVersionResource{
+		Group:    gatewayv1.GroupVersion.Group,
+		Version:  gatewayv1.GroupVersion.Version,
+		Resource: "gateways",
+	}
+	gvrToListKind := map[schema.GroupVersionResource]string{
+		gatewayv1GVR: "GatewayList",
+	}
+	for _, obj := range initRuntimeObjects {
+		if crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition); ok {
+			gvr := schema.GroupVersionResource{
+				Group:    crd.Spec.Group,
+				Version:  crd.Spec.Versions[0].Name,
+				Resource: crd.Spec.Names.Plural, // CRD Kinds directly map to the Resource.
+			}
+			gvrToListKind[gvr] = crd.Spec.Names.Kind + "List"
+		}
+	}
+	fakeDC := fakedynamicclient.NewSimpleDynamicClientWithCustomListKinds(scheme, gvrToListKind)
+	for _, obj := range initRuntimeObjects {
+		var err error
+		if gateway, ok := obj.(*gatewayv1.Gateway); ok {
+			// Register Gateway with correct GVR (since the guessed GVR is incorrect).
+			err = fakeDC.Tracker().Create(gatewayv1GVR, gateway, gateway.Namespace)
```

Comment:
Why do we have a special case for Gateways but no other types?

Reply:
Ah nice question. Sorry I've expanded the comment now:

```
// Register Gateway with correct GVR. This needs to be done explicitly for
// Gateway since the automatically guess GVR is incorrect.
//
// Automatic guessing of GVR uses `meta.UnsafeGuessKindToResource()` which
// pluralizes "gateway" to "gatewaies" (since the singular ends in a 'y')
```